Data/ByteString/Internal/Type.hs

```diff
@@ -354,9 +354,17 @@ instance IsString ByteString where
 
 instance Data ByteString where
   gfoldl f z txt = z packBytes `f` unpackBytes txt
-  toConstr _     = error "Data.ByteString.ByteString.toConstr"
-  gunfold _ _    = error "Data.ByteString.ByteString.gunfold"
-  dataTypeOf _   = mkNoRepType "Data.ByteString.ByteString"
+  toConstr _     = packConstr
```

Comment:
I wasn't sure if this should be strict, but after thinking about it, the choice to be lazy (newtype-like) was already made by `gfoldl` and we should be consistent here.